javascript-style-guide.md

```diff
@@ -143,6 +179,131 @@ algorithms that might grow the call stack boundlessly.
 precise and oftentimes more useful than their procedural alternatives.
 
+## Object-orientation
+
+You can leverage object-oriented design patterns such as encapsulation,
+composition, and inheritance in JavaScript. Create functions whose instance
+variables represent private and protected members. From these functions, return
+objects with methods to access and possibly manipulate the instance
+variables. Create functions with extended interfaces that mutate instances of
+their parents to effect inheritance.
+
+Example:
+
+```js
+var makeMammal = function (data) {
+    var self = {};
+    self.getName = function () {
+        return data.name;
+    };
+    self.says = function () {
+        return data.saying === undefined ? '' : data.saying;
+    };
+    return self;
+};
+
+var myMammal = makeMammal({name: 'Herb'});
+
+var makeCat = function (data) {
+    var self;
+    data.saying = data.saying === undefined ? 'meow' : data.saying;
+    self = makeMammal(data);
+    self.purr = function (n) {
+        var message = '',
+            index;
+        for (index = 0; index < n; index += 1) {
+            if (message.length > 0) {
+                message += '-';
+            }
+            message += 'r';
+        }
+        return message;
+    };
+    self.getName = function () {
+        return self.says() + ' ' + data.name +
+            ' ' + self.says();
+    };
+    return self;
+};
+
+var myCat = makeCat({name: 'Henrietta'});
+```
+
+## Existence
+
+In JavaScript, the values `false`, `undefined` and `null` are falsy, but so are
+`''` and `0`. This is unfortunate, because sometimes empty strings and zero are
+"valid" values from a business perspective. Therefore, when checking for the
+"existence" of a value, it is unreliable to rely on truthiness or
+falsiness. Instead, you should explicitly check for negative values.
+
+Do:
+
+```js
+var number = value === undefined ? 100 : value;
+if (number !== undefined) {
+    // ...
+}
+```
+
+Don't:
+
+```js
+var number = value || 100; // Excludes 0.
+if (number) {
+    // Won't execute if the value is 0.
+}
+```
+
+## Simplicity
+
+It's possible to do lots of things on one line. But that can make your code
+complex and confusing. You should try to do just one thing per line.
+
+Do:
+
+```js
+value = array[index];
+index += 1;
+```
+
+Don't:
+
+```js
+value = array[index++];
+```
+
+## Abstraction
+
+Abstract away complex or roundabout routines to make your intent obvious.
+
+For instance, some people use the value returned from `indexOf` to test for the
+existence of substrings or elements. But it would be better to abstract that
+implication into a `contains` function, as you probably care about existence,
+not indices.
+
+Do:
+
+```js
+if (contains(array, 'value')) {
+    // ...
+}
+if (isInteger(value)) {
+    // ...
+}
+```
+
+Don't:
+
+```js
+if (array.indexOf('value') > -1) {
+    // ...
+}
+if (parseInt(value, 10) !== NaN) {
```

Comment:
This comparison will always evaluate to `true`, as `NaN !== NaN` (strangely enough). Even though this is a "Don't" example, use the `isNaN()` function.